### ISSUE TITLE
Checking CUVID sample of CUDA SDK 6.5 crashes (CUDA GPU video dec, 64-bit)

### ISSUE THREAD

**Issue by oscarbg** [wine-compholio/wine-staging#277]
Hi,
I was checking more or less full CUDA SDK 6.5 samples Windows binaries with wine-staging and have to say I'm impressed almost all work including the ones that use CUBLAS,CUFFT,CURAND etc.. well done! as you say DX interop ones doesn't seem to work but that was expected..
More interestingly I was expecting CUDADecodeGL sample to work as can use CUVID and seems there is a wrapper of CUVID implemented..
note this sample has some modes of operation(decoding) by default it uses CUDA Video decoding (not the same as CUVID (which uses HW video blocks called purevideo) it's using pure CUDA kernels and doesn't work.. to test CUVID you must use "-decodecuvid" arguments and I tried to disabled GL interop with  "-nointerop" but that doesn't helped much.. 
(more details: really seems decodecuvid flag is ignored and had to recompile sample on Windows forcing to using CUVID mode: anyway that doesn't worked too)
have to say I'm interested in CUVID working because in case you don't know it's one of two ways of how HEVC hybrid decoding it's exposed on Windows (in addition to DXVA2 mode).. it's implemented on LAV filters and Media player home cinema has integrated so that way perhaps one day we can play HEVC with some form of HW accel. this way on Linux..

**Comment by slackner** [comment]
Hi,

I just did some tests, and can confirm that the CUDA SDK 6.5 example crashes, however the CUDA SDK 6.0 example runs fine on my machine (GeForce GTX 650, driver version 343.36).

Do you test on MacOS or on Linux? Please double-check that the output below (especially the crash "read access to 0x00000000") is the same, otherwise you're probably experiencing a different issue.

I assume that the issue is caused by an insufficiency of our wrapper, so marking as an bug.

```
$ WINEDEBUG=+nvapi,+nvcuvid,+nvcuda WINEPREFIX=~/.wine-cuda wine cudaDecodeGL.exe -decodecuvid -displayvideo
fixme:winediag:start_process Wine Staging is a testing version containing experimental patches.
fixme:winediag:start_process Please report bugs at http://bugs.wine-staging.com (instead of winehq.org).
trace:nvcuvid:DllMain (0x7ebf0000, 1, 0x1)
trace:nvcuda:DllMain (0x7ebb0000, 1, 0x1)
fixme:heap:RtlSetHeapInformation (nil) 1 (nil) 0 stub
[CUDA/OpenGL Video Decode]
Command Line Arguments:
argv[0] = Z:\nfs\share\wine\cuda-test-6.5\cudaDecodeGL.exe
argv[1] = -decodecuvid
argv[2] = -displayvideo
wine: Unhandled page fault on read access to 0x00000000 at address 0x408930 (thread 0009), starting debugger...
Unhandled exception: page fault on read access to 0x00000000 in 32-bit code (0x00408930).
Register dump:
 CS:0023 SS:002b DS:002b ES:002b FS:0063 GS:006b
 EIP:00408930 ESP:0033fca8 EBP:0033fdb8 EFLAGS:00010206(  R- --  I   - -P- )
 EAX:00000000 EBX:00771740 ECX:9db59b26 EDX:0033fcb4
 ESI:00000000 EDI:00000003
Stack dump:
0x0033fca8:  00771740 7b8b7000 00000003 00000040
0x0033fcb8:  007725f0 007725f0 00000002 00772720
0x0033fcc8:  7bcce000 7bc3df16 0033fd18 f74f0032
0x0033fcd8:  7bcce000 7bc8dd6a 00000004 0033fcf0
0x0033fce8:  0033fd48 7bc513ea 000fe5bd 32b9deb0
0x0033fcf8:  004149d3 7bc51142 004149c9 7bc8dd59
000c: sel=0067 base=00000000 limit=00000000 16-bit rw-
Backtrace:
=>0 0x00408930 in cudadecodegl (+0x8930) (0x0033fdb8)
  1 0x00408a2f in cudadecodegl (+0x8a2e) (0x0033fde8)
  2 0x0040cb07 in cudadecodegl (+0xcb06) (0x0033fe30)
  3 0x7b85c5ec call_process_entry+0xb() in kernel32 (0x0033fe48)
  4 0x7b85d952 start_process+0x91(peb=<couldn't compute location>) [/build/tmp/arch-IuM7hE/wine-staging/src/build-32/dlls/kernel32/../../../wine/dlls/kernel32/process.c:1122] in kernel32 (0x0033fe88)
  5 0x7bc7ede0 call_thread_func_wrapper+0xb() in ntdll (0x0033fea8)
  6 0x7bc81edf call_thread_func+0xce(entry=<couldn't compute location>, arg=<couldn't compute location>, frame=<couldn't compute location>) [/build/tmp/arch-IuM7hE/wine-staging/src/build-32/dlls/ntdll/../../../wine/dlls/ntdll/signal_i386.c:2994] in ntdll (0x0033ff88)
  7 0x7bc7edbe RtlRaiseException+0x21() in ntdll (0x0033ffb8)
  8 0x7bc53491 start_process+0x20(kernel_start=<couldn't compute location>) [/build/tmp/arch-IuM7hE/wine-staging/src/build-32/dlls/ntdll/../../../wine/dlls/ntdll/loader.c:2892] in ntdll (0x0033ffd8)
  9 0xf760699d wine_call_on_stack+0x1c() in libwine.so.1 (0x00000000)
  10 0xf7606b1a wine_switch_to_stack+0x29(func=<couldn't compute location>, arg=<couldn't compute location>, stack=<couldn't compute location>) [/build/tmp/arch-IuM7hE/wine-staging/src/build-32/libs/wine/../../../wine/libs/wine/port.c:59] in libwine.so.1 (0xff8f7948)
  11 0x7bc58bf5 LdrInitializeThunk+0x1f4(kernel_start=<couldn't compute location>, unknown2=<couldn't compute location>, unknown3=<couldn't compute location>, unknown4=<couldn't compute location>) [/build/tmp/arch-IuM7hE/wine-staging/src/build-32/dlls/ntdll/../../../wine/dlls/ntdll/loader.c:2946] in ntdll (0xff8f7998)
  12 0x7b863a83 __wine_kernel_init+0x862() [/build/tmp/arch-IuM7hE/wine-staging/src/build-32/dlls/kernel32/../../../wine/dlls/kernel32/process.c:1294] in kernel32 (0xff8f8a98)
  13 0x7bc59a03 __wine_process_init+0x162() [/build/tmp/arch-IuM7hE/wine-staging/src/build-32/dlls/ntdll/../../../wine/dlls/ntdll/loader.c:3155] in ntdll (0xff8f8b08)
  14 0xf7604553 wine_init+0x2e2(argc=<couldn't compute location>, argv=<couldn't compute location>, error=<couldn't compute location>, error_size=<couldn't compute location>) [/build/tmp/arch-IuM7hE/wine-staging/src/build-32/libs/wine/../../../wine/libs/wine/loader.c:958] in libwine.so.1 (0xff8f8b78)
  15 0x7bf00f42 main+0x81() in <wine-loader> (0xff8f8fd8)
  16 0xf7410e5e __libc_start_main+0xdd() in libc.so.6 (0x00000000)
0x00408930: movb	0x0(%eax),%cl
Modules:
Module	Address			Debug info	Name (51 modules)
PE	  340000-  388000	Deferred        freeglut
PE	  400000-  439000	Export          cudadecodegl
PE	10000000-10033000	Deferred        glew32
ELF	7a800000-7a91d000	Deferred        opengl32<elf>
  \-PE	7a820000-7a91d000	\               opengl32
ELF	7b800000-7ba5e000	Dwarf           kernel32<elf>
  \-PE	7b810000-7ba5e000	\               kernel32
ELF	7bc00000-7bceb000	Dwarf           ntdll<elf>
  \-PE	7bc10000-7bceb000	\               ntdll
ELF	7bf00000-7bf04000	Dwarf           <wine-loader>
ELF	7d2fd000-7e083000	Deferred        libcuda.so
ELF	7e083000-7e341000	Deferred        libnvcuvid.so
ELF	7e34a000-7e36e000	Deferred        imm32<elf>
  \-PE	7e350000-7e36e000	\               imm32
ELF	7e36e000-7e397000	Deferred        libexpat.so.1
ELF	7e397000-7e3d3000	Deferred        libfontconfig.so.1
ELF	7e3d3000-7e446000	Deferred        libpcre.so.1
ELF	7e446000-7e56d000	Deferred        libglib-2.0.so.0
ELF	7e56d000-7e5ca000	Deferred        libharfbuzz.so.0
ELF	7e5ca000-7e609000	Deferred        libpng16.so.16
ELF	7e609000-7e61a000	Deferred        libbz2.so.1.0
ELF	7e61a000-7e631000	Deferred        libz.so.1
ELF	7e631000-7e6f4000	Deferred        libfreetype.so.6
ELF	7e6f4000-7e759000	Deferred        libncursesw.so.5
ELF	7e786000-7e7b1000	Deferred        msacm32<elf>
  \-PE	7e790000-7e7b1000	\               msacm32
ELF	7e7b1000-7e832000	Deferred        rpcrt4<elf>
  \-PE	7e7c0000-7e832000	\               rpcrt4
ELF	7e832000-7e969000	Deferred        ole32<elf>
  \-PE	7e850000-7e969000	\               ole32
ELF	7e969000-7ea20000	Deferred        winmm<elf>
  \-PE	7e970000-7ea20000	\               winmm
ELF	7ea20000-7ea3a000	Deferred        version<elf>
  \-PE	7ea30000-7ea3a000	\               version
ELF	7ea3a000-7eba7000	Deferred        user32<elf>
  \-PE	7ea50000-7eba7000	\               user32
ELF	7eba7000-7ebe0000	Deferred        nvcuda<elf>
  \-PE	7ebb0000-7ebe0000	\               nvcuda
ELF	7ebe0000-7ebf7000	Deferred        nvcuvid<elf>
  \-PE	7ebf0000-7ebf7000	\               nvcuvid
ELF	7ebf7000-7ec67000	Deferred        advapi32<elf>
  \-PE	7ec00000-7ec67000	\               advapi32
ELF	7ec67000-7ed86000	Deferred        gdi32<elf>
  \-PE	7ec70000-7ed86000	\               gdi32
ELF	7ef86000-7efd3000	Deferred        libm.so.6
ELF	f73f4000-f73f9000	Deferred        libdl.so.2
ELF	f73f9000-f75b1000	Dwarf           libc.so.6
ELF	f75b1000-f75ce000	Deferred        libpthread.so.0
ELF	f75fb000-f77bf000	Dwarf           libwine.so.1
ELF	f77c0000-f77e3000	Deferred        ld-linux.so.2
ELF	f77e5000-f77e6000	Deferred        [vdso].so
Threads:
process  tid      prio (all id:s are in hex)
00000008 (D) Z:\nfs\share\wine\cuda-test-6.5\cudaDecodeGL.exe
	00000009    0 <==
0000000e services.exe
	0000001d    0
	0000001c    0
	00000016    0
	00000014    0
	00000010    0
	0000000f    0
00000012 winedevice.exe
	0000001b    0
	00000018    0
	00000017    0
	00000013    0
00000019 plugplay.exe
	0000001f    0
	0000001e    0
	0000001a    0
00000020 explorer.exe
	00000021    0
```

Regards,
Sebastian

**Comment by slackner** [comment]
Oh, i have to take back all what I said above. It seems like I didn't have the video data files for CUDA 6.5 SDK extracted properly. After doing that and it works fine here. (It might be necessary to switch working directory to where the cudaDecodeGL.exe executeable is located).

Please post some terminal output from your machine.

**Comment by oscarbg** [comment]
Some update with good and bad news (briefly win32 binary works but I was checking win64 binary and that doesn't work):
my system is ubuntu 14.10, drv 346.22 and gtx 970
I checked what you said about launching from correct working directory: it was OK already..
I was trying to launch 64bit windows executable sample (from CUDASamples/bin/win64/Release) that is still not working.. see output:
> VideoDecoder::cudaVideoCreateFlags = <4>Use CUVID decoder
trace:nvcuvid:wine_cuvidCreateDecoder (0x59df0, 0x59d80)
Assertion failed: CUDA_SUCCESS == oResult, file VideoDecoder.cpp, line 102
wine: Unhandled exception 0x40000015 in thread 36 at address 0x1400108ce (thread 0036), starting debugger...
Good news is 32bit binary is working so you must be testing win32 binary right?
Still is a bug as other 64 bit Windows binaries from CUDA SDK work right!
Also please check that you see:
<4>Use CUVID decoder
as me even using -decodecuvid flag was seeing something like
<1>Use CUDA decoder..

**Comment by slackner** [comment]
For me the output looks a bit different, but I can confirm that there are still some issues with nvcuvid on x86_64. Will take closer look.

**Comment by slackner** [comment]
Should be fixed with d50c67314440ceed7510abce922d2670f938ae36.

With those fixes the CUDA SDK demo works fine in a 64-bit environment for me - can you confirm that?

**Comment by slackner** [comment]
Assuming fixed. If you still encounter any remaining issues after the next release please open a new bug report (preferably in our [new bugtracker](http://bugs.wine-staging.com)).
